pyro/contrib/epidemiology/compartmental.py

```diff
@@ -0,0 +1,427 @@
+                                    .to_event(2))
+
+        # Sequentially transition.
+        curr = self.initialize(params)
+        for t in poutine.markov(range(self.duration)):
+            aux_t = auxiliary[..., t]
+            prev = curr
+            curr = {name: quantize("{}_{}".format(name, t), aux,
+                                   min=0, max=self.population)
+                    for name, aux in zip(self.compartments, aux_t.unbind(-1))}
+            self.transition_bwd(params, prev, curr, t)
+
+    def _vectorized_model(self):
+        """
+        Vectorized model used for inference.
+        """
+        # Sample global parameters.
+        params = self.global_model()
+
+        # Sample the continuous reparameterizing variable.
+        auxiliary = pyro.sample("auxiliary",
+                                dist.Uniform(-0.5, self.population + 0.5)
+                                    .mask(False)
+                                    .expand([len(self.compartments), self.duration])
+                                    .to_event(2))
+
+        # Manually enumerate.
+        curr, logp = quantize_enumerate(auxiliary, min=0, max=self.population)
+        curr = OrderedDict(zip(self.compartments, curr))
+        logp = OrderedDict(zip(self.compartments, logp))
+
+        # Truncate final value from the right then pad initial value onto the left.
+        init = self.initialize(params)
+        prev = {}
+        for name in self.compartments:
+            if not isinstance(init[name], int):
+                raise NotImplementedError("TODO use torch.cat()")
+            prev[name] = pad(curr[name][:-1], (0, 0, 1, 0), value=init[name])
+
+        # Reshape to support broadcasting, similar to EnumMessenger.
+        C = len(self.compartments)
+        T = self.duration
+        Q = 4  # Number of quantization points.
+
+        def enum_shape(position):
+            shape = [T] + [1] * (2 * C)
+            shape[1 + position] = Q
+            return torch.Size(shape)
+
+        for e, name in enumerate(self.compartments):
+            prev[name] = prev[name].reshape(enum_shape(e))
+            curr[name] = curr[name].reshape(enum_shape(C + e))
+            logp[name] = logp[name].reshape(enum_shape(C + e))
+        t = (Ellipsis,) + (None,) * (2 * C)  # Used to unsqueeze data tensors.
+
+        # Record transition factors.
+        with poutine.block(), poutine.trace() as tr:
+            self.transition_bwd(params, prev, curr, t)
+        for name, site in tr.trace.nodes.items():
+            if site["type"] == "sample":
+                logp[name] = site["fn"].log_prob(site["value"])
+
+        # Manually perform variable elimination.
+        logp = sum(logp.values())
+        logp = logp.reshape(T, Q ** C, Q ** C)
+        logp = pyro.distributions.hmm._sequential_logmatmulexp(logp)
+        logp = logp.reshape(-1).logsumexp(0)
+        warn_if_nan(logp)
+        pyro.factor("transition", logp)
```

Comment:
what happened to the previous `-log(4)`?

Reply:
I have neglected that from this model because (1) it does not affect inference (it does not change gradients), and (2) we will probably be moving to stochastic initial state soon anyway.